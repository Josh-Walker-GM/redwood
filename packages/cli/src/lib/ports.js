import portfinder from 'portfinder'

/**
 * Finds a free port
 * @param  {[number]}   requestedPort Port to start searching from
 * @param  {[number[]]} excludePorts  Array of port numbers to exclude
 * @return {[number]}                 A free port equal or higher than requestedPort but not within excludePorts. If no port can be found then returns -1
 */
export async function getFreePort(requestedPort, excludePorts = []) {
  try {
    let freePort = await portfinder.getPortPromise({
      port: requestedPort,
    })
    if (excludePorts.includes(freePort)) {
      freePort = await getFreePort(freePort + 1, excludePorts)
    }
    return freePort
  } catch (error) {
    return -1
  }
}

/**
 * Determines if a port is available or in use
 * @param {number} port The port number to check
 * @return {boolean} True if the port is available, false otherwise
 */
export async function isPortFree(port) {
  return (await portfinder.getPortPromise({ port })) === port
}
