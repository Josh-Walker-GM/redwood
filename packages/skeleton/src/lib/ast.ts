import fs from 'fs'

import { parse } from '@babel/parser'
import traverse from '@babel/traverse'
import {
  assertJSXOpeningElement,
  assertJSXIdentifier,
  assertIdentifier,
  isJSXAttribute,
  isJSXSpreadAttribute,
  JSXElement,
  ExportNamedDeclaration,
  isVariableDeclaration,
  isVariableDeclarator,
  isIdentifier,
  File,
  VariableDeclarator,
  ExportDefaultDeclaration,
} from '@babel/types'

export function getASTFromCode(code: string) {
  return parse(code, {
    sourceType: 'unambiguous',
    // TODO: Check these plugin options are optimal, I doubt they are
    plugins: [
      'jsx',
      'typescript',
      'nullishCoalescingOperator',
      'objectRestSpread',
    ],
  })
}

export function getASTFromFile(filePath: string) {
  const code = fs.readFileSync(filePath, { encoding: 'utf8', flag: 'r' })
  return getASTFromCode(code)
}

export function getSpecificNamedExportDeclarations(ast: File, names: string[]) {
  const desiredNamedExports: Map<string, VariableDeclarator> = new Map()
  const allNamedExports: ExportNamedDeclaration[] = []
  traverse(ast, {
    ExportNamedDeclaration: (path) => {
      allNamedExports.push(path.node)
    },
  })
  allNamedExports.forEach((namedExport) => {
    if (
      namedExport.declaration != null &&
      isVariableDeclaration(namedExport.declaration) &&
      isVariableDeclarator(namedExport.declaration.declarations[0]) &&
      isIdentifier(namedExport.declaration.declarations[0].id) &&
      names.includes(namedExport.declaration.declarations[0].id.name)
    ) {
      desiredNamedExports.set(
        namedExport.declaration.declarations[0].id.name,
        namedExport.declaration.declarations[0]
      )
    }
  })
  return desiredNamedExports
}

export function getExportDefaultDeclaration(ast: File) {
  let defaultExport: ExportDefaultDeclaration | undefined
  traverse(ast, {
    ExportDefaultDeclaration: (path) => {
      defaultExport = path.node
    },
  })
  return defaultExport
}

export function getVariableDeclarator(ast: File, name: string) {
  let variableDeclarator: VariableDeclarator | undefined
  traverse(ast, {
    VariableDeclarator: (path) => {
      if (isIdentifier(path.node.id) && path.node.id.name === name) {
        variableDeclarator = path.node
      }
    },
  })
  return variableDeclarator
}

export function getJSXElementAttributes(
  element: JSXElement
): Map<string, string> {
  const attributes: Map<string, string> = new Map()
  assertJSXOpeningElement(element.openingElement)
  if (element.openingElement.attributes.length > 0) {
    element.openingElement.attributes.forEach((attribute) => {
      if (isJSXAttribute(attribute)) {
        assertJSXIdentifier(attribute.name)
        const key = attribute.name.name
        let value: string
        switch (attribute.value?.type) {
          case 'StringLiteral':
            value = attribute.value.value
            break
          case 'JSXExpressionContainer':
            assertIdentifier(attribute.value.expression)
            value = attribute.value.expression.name
            break
          case undefined: // Accept this because it could be a attribute with no value, i.e. notfound
            value = 'true'
            break
          default:
            throw new Error(
              `Unknown value type found while trying to process the ${key} property when extracting all attributes`
            )
        }
        attributes.set(key, value)
      } else if (isJSXSpreadAttribute(attribute)) {
        throw new Error(
          'JSXSpreadAttribute type is not yet supported when extracting all attributes'
        )
      } else {
        throw new Error(
          'Uknown attribute type found while trying to extract all attributes'
        )
      }
    })
  }
  return attributes
}
