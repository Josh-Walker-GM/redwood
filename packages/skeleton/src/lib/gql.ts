import { parse } from 'graphql'

export function parseGraphQL(gql: string) {
  return parse(gql)
}

export function getGraphQLQueryName(gql: string): string | undefined {
  const ast = parse(gql)
  for (const def of ast.definitions) {
    if (def.kind === 'OperationDefinition' && def.operation === 'query') {
      return def.name?.value
    }
  }
  return undefined
}
