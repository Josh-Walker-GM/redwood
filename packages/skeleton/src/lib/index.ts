import fs from 'fs'

import { parse } from '@babel/parser'

export function getFileAST(filePath: string) {
  const code = fs.readFileSync(filePath, { encoding: 'utf8', flag: 'r' })
  return parse(code, {
    sourceType: 'unambiguous',
    plugins: ['jsx', 'typescript'],
  })
}
