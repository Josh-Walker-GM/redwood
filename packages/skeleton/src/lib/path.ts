import path from 'path'

import { getBaseDir, getConfigPath } from '@redwoodjs/internal/dist/index'

/**
 * Determines the root path of a redwood project which contains the provided path
 * @param pathWithinProject A path within a redwood project
 * @returns The root path of the associated project
 */
export function getRootPath(pathWithinProject: string) {
  return pathWithinProject
    ? path.dirname(getConfigPath(pathWithinProject))
    : getBaseDir()
}
