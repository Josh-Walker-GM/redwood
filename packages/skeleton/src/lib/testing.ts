import { ensurePosixPath } from '@redwoodjs/internal/dist/paths'

export function stripAndFormatPathForTesting(
  filepath: string,
  projectFilepath: string
) {
  if (filepath.startsWith(projectFilepath)) {
    filepath = ensurePosixPath(filepath.substring(projectFilepath.length))
  }
  return filepath
}
