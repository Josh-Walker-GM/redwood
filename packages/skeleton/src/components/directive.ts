import fs from 'fs'
import path from 'path'

import traverse from '@babel/traverse'
import {
  isCallExpression,
  isIdentifier,
  isTemplateElement,
  VariableDeclarator,
} from '@babel/types'
import { DirectiveDefinitionNode } from 'graphql'

import { getPaths } from '@redwoodjs/internal/dist/paths'

import {
  getASTFromFile,
  getExportDefaultDeclaration,
  getSpecificNamedExportDeclarations,
} from '../lib/ast'
import { parseGraphQL } from '../lib/gql'

import {
  RedwoodError,
  RedwoodErrorCode,
  RedwoodWarning,
  RedwoodWarningCode,
} from './diagnostic'
import type { RedwoodProject } from './project'
import { RedwoodSkeleton } from './skeleton'

export class RedwoodDirective extends RedwoodSkeleton {
  warnings: RedwoodWarning[] = []
  errors: RedwoodError[] = []

  readonly gql: string | undefined
  readonly kind: 'transform' | 'validate' | 'unknown' = 'unknown'

  constructor(filepath: string) {
    const warnings: RedwoodWarning[] = []
    const errors: RedwoodError[] = []
    let name: string | undefined
    let gql: string | undefined

    const ast = getASTFromFile(filepath)

    const requiredNamedExports = getSpecificNamedExportDeclarations(ast, [
      'schema',
    ])

    if (requiredNamedExports.has('schema')) {
      const schemaExportVariableDeclarator = requiredNamedExports.get(
        'schema'
      ) as VariableDeclarator
      switch (schemaExportVariableDeclarator.init?.type) {
        case 'TaggedTemplateExpression':
          if (
            schemaExportVariableDeclarator.init.quasi.quasis[0] !== undefined &&
            isTemplateElement(
              schemaExportVariableDeclarator.init.quasi.quasis[0]
            )
          ) {
            gql =
              schemaExportVariableDeclarator.init.quasi.quasis[0].value.cooked
          }
          break

        default:
          warnings.push({
            code: RedwoodWarningCode.GENERIC_PARSER_WARNING_JSTS,
            message: `Unable to extract the gql (${schemaExportVariableDeclarator.init?.type})`,
          })
          break
      }
      if (gql === undefined) {
        warnings.push({
          code: RedwoodWarningCode.GENERIC_PARSER_WARNING_JSTS,
          message: 'Could not extract the gql',
        })
      } else {
        const gqlAST = parseGraphQL(gql)
        const directiveDefinitions = gqlAST.definitions.filter(
          (node): node is DirectiveDefinitionNode => {
            return node.kind === 'DirectiveDefinition'
          }
        )
        if (directiveDefinitions.length !== 1) {
          errors.push({
            code: RedwoodErrorCode.DIRECTIVE_DEFINE_ONE_DIRECTIVE,
            message: 'Must define one directive',
          })
        } else {
          name = directiveDefinitions[0].name.value
        }
      }
    } else {
      errors.push({
        code: RedwoodErrorCode.DIRECTIVE_MISSING_SCHEMA_EXPORT,
        message: 'No "schema" export could be found',
      })
    }

    super(filepath, name)
    this.gql = gql
    this.errors = errors
    this.warnings = warnings

    const defaultExport = getExportDefaultDeclaration(ast)
    if (defaultExport !== undefined) {
      let directiveHelperFunctionName
      if (isIdentifier(defaultExport.declaration)) {
        const defaultExportVariable = defaultExport.declaration.name
        traverse(ast, {
          VariableDeclarator: (path) => {
            if (
              isIdentifier(path.node.id) &&
              path.node.id.name === defaultExportVariable &&
              isCallExpression(path.node.init) &&
              isIdentifier(path.node.init.callee)
            ) {
              directiveHelperFunctionName = path.node.init.callee.name
            }
          },
        })
      } else if (
        isCallExpression(defaultExport.declaration) &&
        isIdentifier(defaultExport.declaration.callee)
      ) {
        directiveHelperFunctionName = defaultExport.declaration.callee.name
      } else {
        this.warnings.push({
          code: RedwoodWarningCode.GENERIC_PARSER_WARNING_JSTS,
          message: 'The default export could not be understood',
        })
      }

      switch (directiveHelperFunctionName) {
        case 'createValidatorDirective':
          this.kind = 'validate'
          break
        case 'createTransformerDirective':
          this.kind = 'transform'
          break
        case undefined:
          this.warnings.push({
            code: RedwoodWarningCode.GENERIC_PARSER_WARNING_JSTS,
            message: `Could not understand the directive function`,
          })
          break
        default:
          this.warnings.push({
            code: RedwoodWarningCode.GENERIC_PARSER_WARNING_JSTS,
            message: `Could not understand the directive function "${directiveHelperFunctionName}"`,
          })
          break
      }
    } else {
      this.errors.push({
        code: RedwoodErrorCode.DIRECTIVE_MISSING_DEFAULT_EXPORT,
        message: 'Must have a default export which is the directive function',
      })
    }
  }
}

export function extractDirective(filepath: string): RedwoodDirective {
  return new RedwoodDirective(filepath)
}

export function extractDirectives(
  project: RedwoodProject | undefined = undefined
): RedwoodDirective[] {
  const directives: RedwoodDirective[] = []

  const directivesPath = project
    ? getPaths(project.filepath).api.directives
    : getPaths().api.directives

  if (!fs.existsSync(directivesPath)) {
    return directives
  }

  // TODO: Confirm this is the condition to detect a directive
  const getDirectiveFiles = (directory: string) => {
    const directiveFiles: string[] = []
    const directoryContents = fs.readdirSync(directory)
    directoryContents.forEach((content) => {
      const stat = fs.lstatSync(path.join(directory, content))
      if (stat.isDirectory()) {
        directiveFiles.push(...getDirectiveFiles(path.join(directory, content)))
      } else if (stat.isFile()) {
        if (
          content.match(/.+\.(js|ts)$/) &&
          !content.match(/.+\.(test\.js|test\.ts)$/)
        ) {
          directiveFiles.push(path.join(directory, content))
        }
      }
    })
    return directiveFiles
  }

  const directiveFiles = getDirectiveFiles(directivesPath)
  directiveFiles.forEach((directivePath) => {
    directives.push(extractDirective(directivePath))
  })

  return directives
}
