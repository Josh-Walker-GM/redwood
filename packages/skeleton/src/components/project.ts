import fs from 'fs'
import path from 'path'

import chalk from 'chalk'

import { getPaths } from '@redwoodjs/internal/dist/index'

import { getRootPath } from '../lib/path'

import { RedwoodSkeleton } from './base'
import { extractCells, RedwoodCell } from './cell'
import { extractFunctions, RedwoodFunction } from './function'
import { extractLayouts, RedwoodLayout } from './layout'
import { extractPages, RedwoodPage } from './page'
import { extractRouters, RedwoodRouter } from './router'
import { extractSides, RedwoodSide } from './side'

/**
 * Used to enumerate either JS or TS project types
 */
export enum RedwoodProjectType {
  TYPESCRIPT = 'typescript',
  JAVASCRIPT = 'javascript',
}

export class RedwoodProject extends RedwoodSkeleton {
  warnings: string[] = []
  errors: string[] = []

  readonly type: RedwoodProjectType

  private sides?: RedwoodSide[] | undefined
  private cells?: RedwoodCell[] | undefined
  private routers?: RedwoodRouter[] | undefined
  private layouts?: RedwoodLayout[] | undefined
  private pages?: RedwoodPage[] | undefined
  private functions?: RedwoodFunction[] | undefined

  public static getProject({
    pathWithinProject = '',
    full = false,
    readFromCache = true,
    insertIntoCache = true,
  }: {
    pathWithinProject?: string
    full?: boolean
    readFromCache?: boolean
    insertIntoCache?: boolean
  } = {}) {
    const projectCache = RedwoodProjectsCache.getInstance().projects
    const rootPath = getRootPath(pathWithinProject)

    if (readFromCache) {
      const cachedProject = projectCache.get(rootPath)
      if (cachedProject) {
        return cachedProject
      }
    }

    const project = new RedwoodProject(rootPath, full)
    if (insertIntoCache) {
      projectCache.set(project.filepath, project)
    }
    return project
  }

  private constructor(rootPath: string, full = false) {
    super(rootPath)

    // A project is typescript if we detect a tsconfig.json
    this.type =
      fs.existsSync(
        path.join(getPaths(this.filepath).web.base, 'tsconfig.json')
      ) ||
      fs.existsSync(
        path.join(getPaths(this.filepath).api.base, 'tsconfig.json')
      )
        ? RedwoodProjectType.TYPESCRIPT
        : RedwoodProjectType.JAVASCRIPT

    if (full) {
      this.sides = extractSides(this)
      this.cells = extractCells(this)
      this.layouts = extractLayouts(this)
      this.pages = extractPages(this)
      this.functions = extractFunctions(this)
      this.routers = extractRouters(this)
    }
  }

  getSides(forceExtract = false): RedwoodSide[] {
    if (forceExtract) {
      return extractSides(this)
    }
    return this.sides === undefined ? extractSides(this) : this.sides
  }

  getCells(forceExtract = false): RedwoodCell[] {
    if (forceExtract) {
      return extractCells(this)
    }
    return this.cells === undefined ? extractCells(this) : this.cells
  }

  getRouters(forceExtract = false): RedwoodRouter[] {
    if (forceExtract) {
      return extractRouters(this)
    }
    return this.routers === undefined ? extractRouters(this) : this.routers
  }

  getLayouts(forceExtract = false): RedwoodLayout[] {
    if (forceExtract) {
      return extractLayouts(this)
    }
    return this.layouts === undefined ? extractLayouts(this) : this.layouts
  }

  getPages(forceExtract = false): RedwoodPage[] {
    if (forceExtract) {
      return extractPages(this)
    }
    return this.pages === undefined ? extractPages(this) : this.pages
  }

  getFunctions(forceExtract = false): RedwoodFunction[] {
    if (forceExtract) {
      return extractFunctions(this)
    }
    return this.functions === undefined
      ? extractFunctions(this)
      : this.functions
  }

  // Diagnostics

  getInformation(): string {
    return '' // TODO: Implement, gets the info of all other components too
  }

  printInformation(): void {
    console.log(
      `${chalk.bgCyan('[Info]')}\t${this.name}\t${chalk.dim(this.filepath)}`
    )
  }

  printWarnings(cascade = false): void {
    if (this.warnings.length > 0) {
      const titleLine = `${chalk.bgYellow('[Warn]')}\t${this.name} ${chalk.dim(
        this.filepath
      )}`
      const warningLines = this.warnings.map((warning, index) => {
        return ` (${index + 1}) ${warning}\n`
      })
      console.log(titleLine.concat('\n', ...warningLines).trimEnd())
    }
    if (cascade) {
      this.cells?.forEach((cell) => {
        cell.printWarnings()
      })
      this.functions?.forEach((func) => {
        func.printWarnings()
      })
      this.layouts?.forEach((layout) => {
        layout.printWarnings()
      })
      this.pages?.forEach((page) => {
        page.printWarnings()
      })
      this.routers?.forEach((router) => {
        router.printWarnings()
        router.routes.forEach((route) => {
          route.printWarnings()
        })
      })
      this.sides?.forEach((side) => {
        side.printWarnings()
      })
    }
  }

  printErrors(cascade = false): void {
    if (this.errors.length > 0) {
      const titleLine = `${chalk.bgRed('[Error]')}\t${this.name} ${chalk.dim(
        this.filepath
      )}`
      const errorLines = this.errors.map((error, index) => {
        return ` (${index + 1}) ${error}\n`
      })
      console.log(titleLine.concat('\n', ...errorLines).trimEnd())
    }
    if (cascade) {
      this.cells?.forEach((cell) => {
        cell.printErrors()
      })
      this.functions?.forEach((func) => {
        func.printErrors()
      })
      this.layouts?.forEach((layout) => {
        layout.printErrors()
      })
      this.pages?.forEach((page) => {
        page.printErrors()
      })
      this.routers?.forEach((router) => {
        router.printErrors()
        router.routes.forEach((route) => {
          route.printErrors()
        })
      })
      this.sides?.forEach((side) => {
        side.printErrors()
      })
    }
  }
}

/**
 * Singelton class which contains a cache of {@link RedwoodProject} instances.
 *
 * This class is not exported and should only be available from static methods within {@link RedwoodProject}, in order to keep the surface of the skeleton API cleaner and leaner.
 */
class RedwoodProjectsCache {
  private static instance: RedwoodProjectsCache

  public projects: Map<string, RedwoodProject>

  private constructor() {
    this.projects = new Map()
  }

  public static getInstance(): RedwoodProjectsCache {
    if (!RedwoodProjectsCache.instance) {
      RedwoodProjectsCache.instance = new RedwoodProjectsCache()
    }
    return RedwoodProjectsCache.instance
  }
}
