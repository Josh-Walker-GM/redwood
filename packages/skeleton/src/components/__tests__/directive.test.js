import path from 'path'

import { stripAndFormatPathForTesting } from '../../lib/testing'
import { extractDirectives } from '../directive'
import { RedwoodProject } from '../project'

const FIXTURE_PATH = path.join(
  __dirname,
  '..',
  '..',
  '..',
  '..',
  '..',
  '__fixtures__'
)

describe.each([
  'empty-project',
  'example-todo-main',
  'example-todo-main-with-errors',
  'test-project',
])('From within the %s fixture', (PROJECT_NAME) => {
  const PROJECT_PATH = path.join(FIXTURE_PATH, PROJECT_NAME)

  const RWJS_CWD = process.env.RWJS_CWD
  beforeAll(() => {
    process.env.RWJS_CWD = PROJECT_PATH
  })
  afterAll(() => {
    process.env.RWJS_CWD = RWJS_CWD
  })

  it('returns the correct directives without a project', () => {
    const directives = extractDirectives(undefined)
    directives.forEach((directive) => {
      directive.filepath = stripAndFormatPathForTesting(
        directive.filepath,
        PROJECT_PATH
      )
      expect(directive).toMatchSnapshot(directive.filepath)
    })
    expect(directives.length).toMatchSnapshot('directive count')
  })

  it('returns the correct directives with a project', () => {
    const project = RedwoodProject.getProject({
      pathWithinProject: PROJECT_PATH,
      readFromCache: false,
      insertIntoCache: false,
    })
    const directives = extractDirectives(project)
    directives.forEach((directive) => {
      directive.filepath = stripAndFormatPathForTesting(
        directive.filepath,
        PROJECT_PATH
      )
      expect(directive).toMatchSnapshot(directive.filepath)
    })
    expect(directives.length).toMatchSnapshot('directive count')
  })
})
