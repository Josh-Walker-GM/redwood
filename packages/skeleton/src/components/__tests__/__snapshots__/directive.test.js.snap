// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`From within the empty-project fixture returns the correct directives with a project: /api/src/directives/requireAuth/requireAuth.ts 1`] = `
RedwoodDirective {
  "errors": [],
  "filepath": "/api/src/directives/requireAuth/requireAuth.ts",
  "gql": "
  """
  Use to check whether or not a user is authenticated and is associated
  with an optional set of roles.
  """
  directive @requireAuth(roles: [String]) on FIELD_DEFINITION
",
  "kind": "validate",
  "name": "requireAuth",
  "warnings": [],
}
`;

exports[`From within the empty-project fixture returns the correct directives with a project: /api/src/directives/skipAuth/skipAuth.ts 1`] = `
RedwoodDirective {
  "errors": [],
  "filepath": "/api/src/directives/skipAuth/skipAuth.ts",
  "gql": "
  """
  Use to skip authentication checks and allow public access.
  """
  directive @skipAuth on FIELD_DEFINITION
",
  "kind": "validate",
  "name": "skipAuth",
  "warnings": [],
}
`;

exports[`From within the empty-project fixture returns the correct directives with a project: directive count 1`] = `2`;

exports[`From within the empty-project fixture returns the correct directives without a project: /api/src/directives/requireAuth/requireAuth.ts 1`] = `
RedwoodDirective {
  "errors": [],
  "filepath": "/api/src/directives/requireAuth/requireAuth.ts",
  "gql": "
  """
  Use to check whether or not a user is authenticated and is associated
  with an optional set of roles.
  """
  directive @requireAuth(roles: [String]) on FIELD_DEFINITION
",
  "kind": "validate",
  "name": "requireAuth",
  "warnings": [],
}
`;

exports[`From within the empty-project fixture returns the correct directives without a project: /api/src/directives/skipAuth/skipAuth.ts 1`] = `
RedwoodDirective {
  "errors": [],
  "filepath": "/api/src/directives/skipAuth/skipAuth.ts",
  "gql": "
  """
  Use to skip authentication checks and allow public access.
  """
  directive @skipAuth on FIELD_DEFINITION
",
  "kind": "validate",
  "name": "skipAuth",
  "warnings": [],
}
`;

exports[`From within the empty-project fixture returns the correct directives without a project: directive count 1`] = `2`;

exports[`From within the example-todo-main fixture returns the correct directives with a project: /api/src/directives/requireAuth/requireAuth.js 1`] = `
RedwoodDirective {
  "errors": [],
  "filepath": "/api/src/directives/requireAuth/requireAuth.js",
  "gql": "
  directive @requireAuth(roles: [String]) on FIELD_DEFINITION
",
  "kind": "validate",
  "name": "requireAuth",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture returns the correct directives with a project: /api/src/directives/skipAuth/skipAuth.js 1`] = `
RedwoodDirective {
  "errors": [],
  "filepath": "/api/src/directives/skipAuth/skipAuth.js",
  "gql": "
  directive @skipAuth on FIELD_DEFINITION
",
  "kind": "validate",
  "name": "skipAuth",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture returns the correct directives with a project: directive count 1`] = `2`;

exports[`From within the example-todo-main fixture returns the correct directives without a project: /api/src/directives/requireAuth/requireAuth.js 1`] = `
RedwoodDirective {
  "errors": [],
  "filepath": "/api/src/directives/requireAuth/requireAuth.js",
  "gql": "
  directive @requireAuth(roles: [String]) on FIELD_DEFINITION
",
  "kind": "validate",
  "name": "requireAuth",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture returns the correct directives without a project: /api/src/directives/skipAuth/skipAuth.js 1`] = `
RedwoodDirective {
  "errors": [],
  "filepath": "/api/src/directives/skipAuth/skipAuth.js",
  "gql": "
  directive @skipAuth on FIELD_DEFINITION
",
  "kind": "validate",
  "name": "skipAuth",
  "warnings": [],
}
`;

exports[`From within the example-todo-main fixture returns the correct directives without a project: directive count 1`] = `2`;

exports[`From within the example-todo-main-with-errors fixture returns the correct directives with a project: directive count 1`] = `0`;

exports[`From within the example-todo-main-with-errors fixture returns the correct directives without a project: directive count 1`] = `0`;

exports[`From within the test-project fixture returns the correct directives with a project: /api/src/directives/requireAuth/requireAuth.ts 1`] = `
RedwoodDirective {
  "errors": [],
  "filepath": "/api/src/directives/requireAuth/requireAuth.ts",
  "gql": "
  """
  Use to check whether or not a user is authenticated and is associated
  with an optional set of roles.
  """
  directive @requireAuth(roles: [String]) on FIELD_DEFINITION
",
  "kind": "validate",
  "name": "requireAuth",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct directives with a project: /api/src/directives/skipAuth/skipAuth.ts 1`] = `
RedwoodDirective {
  "errors": [],
  "filepath": "/api/src/directives/skipAuth/skipAuth.ts",
  "gql": "
  """
  Use to skip authentication checks and allow public access.
  """
  directive @skipAuth on FIELD_DEFINITION
",
  "kind": "validate",
  "name": "skipAuth",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct directives with a project: directive count 1`] = `2`;

exports[`From within the test-project fixture returns the correct directives without a project: /api/src/directives/requireAuth/requireAuth.ts 1`] = `
RedwoodDirective {
  "errors": [],
  "filepath": "/api/src/directives/requireAuth/requireAuth.ts",
  "gql": "
  """
  Use to check whether or not a user is authenticated and is associated
  with an optional set of roles.
  """
  directive @requireAuth(roles: [String]) on FIELD_DEFINITION
",
  "kind": "validate",
  "name": "requireAuth",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct directives without a project: /api/src/directives/skipAuth/skipAuth.ts 1`] = `
RedwoodDirective {
  "errors": [],
  "filepath": "/api/src/directives/skipAuth/skipAuth.ts",
  "gql": "
  """
  Use to skip authentication checks and allow public access.
  """
  directive @skipAuth on FIELD_DEFINITION
",
  "kind": "validate",
  "name": "skipAuth",
  "warnings": [],
}
`;

exports[`From within the test-project fixture returns the correct directives without a project: directive count 1`] = `2`;
